tests/PHPStan/Analyser/LegacyNodeScopeResolverTest.php

```diff
@@ -3100,13 +3100,21 @@ public function dataBinaryOperations(): array
 				'$coalesceArray',
 			],
 			[
-				'array<0|1|2, 1|2|3>',
+				'array{1, 2, 3}',
```

Comment:
Sorry for lack of explanation. I thought that maybe this test should be kept, because of https://3v4l.org/AitVn

Reply:
I don't know if that's the good direction to go.

This PR is consistent with the current behavior, see https://phpstan.org/r/cd2147fb-6b73-4bdf-abd6-d1a05b0b8334
If we're following what you're asking here, we should report that
```
\PHPStan\dumpType($array[$foo]);
```
is `int|bool|string` then.

IMHO,
When you provide a constant string, phpstan understand `unset($array['1'])` so it's ok.
When you provide a non-constant string, phpstan should report `unset($array[$string])` as useless instead.

It would be consistent with the fact that `$array[$string]` is an error when the array is `array<int, mixed>`.

Reply:
Thank you for the example! Agreed with your explanation.

Reply:
In the original PR I was unsetting everything (for int and string).
This is the current behavior of psalm, see
- https://psalm.dev/r/1296eb7185
- https://psalm.dev/r/be28362a63

@rvanvelzen made this proposal https://github.com/phpstan/phpstan-src/pull/1927#discussion_r1006664191 to only set the key of the same type ; which was not a bad idea.

We can still decide that
- Unsetting a `int` key will only unset the int keys
- Unsetting a `string` key will unset all the possibly keys

That's not a big change.

WDYT @ondrejmirtes ?